File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      Regex.run(~r/^[\p{Lu}' ]*\?[ ]*$/u, input) != nil ->
        # yelled questions
        "Calm down, I know what I'm doing!"

      Regex.run(~r/^(?=.*\p{Lu}{1,})[\p{Lu}0-9![:punct:] ]*$/u, input) != nil ->
        # shouting
        "Whoa, chill out!"

      Regex.run(~r/^[\p{L}[:punct:]0-9 ]*\?[ ]*$/u, input) != nil ->
        # questions
        "Sure."

      Regex.run(~r/^[\n\r \t]*$/u, input) != nil ->
        # Exclamation
        "Fine. Be that way!"

      true ->
        "Whatever."
    end
  end
end

# He answers 'Whoa, chill out!' if you yell at him.

# He answers 'Calm down, I know what I'm doing!' if you yell a question at him.

# He says 'Fine. Be that way!' if you address him without actually saying anything.

# He answers 'Whatever.' to anything else.

File: solutions/erlang/bob/1/src/bob.erl
-module(bob).

-export([response/1]).

response(String) ->
    Trimmed = string:trim(String),
    Upcase = string:uppercase(String),
    Locase = string:lowercase(String),
    Yelled = (Upcase == String) and (Locase /= String),
    Question =
        length(Trimmed) andalso string:slice(Trimmed, string:length(Trimmed) - 1, 1) == "?",
    response(String, Trimmed, Upcase, Locase, Yelled, Question).

response(_String, _Trimmed, _Upcase, _Locase, true, true) ->
    "Calm down, I know what I'm doing!";
response(_String, _Trimmed, _Upcase, _Locase, true, _Question) ->
    "Whoa, chill out!";
response(_String, _Trimmed, _Upcase, _Locase, _Yelled, true) ->
    "Sure.";
response(_String, "", _Upcase, _Locase, _Yelled, _Question) ->
    "Fine. Be that way!";
response(_String, _Trimmed, _Upcase, _Locase, _Yelled, _Question) ->
    "Whatever.".
